delta/service/src/main/scala/ch/epfl/bluebrain/nexus/delta/service/projects/ProjectProvisioning.scala

```diff
@@ -0,0 +1,81 @@
+package ch.epfl.bluebrain.nexus.delta.service.projects
+
+import ch.epfl.bluebrain.nexus.delta.sdk.model.Label
+import ch.epfl.bluebrain.nexus.delta.sdk.model.acls.{Acl, AclAddress, AclRejection}
+import ch.epfl.bluebrain.nexus.delta.sdk.model.identities.Identity.{Subject, User}
+import ch.epfl.bluebrain.nexus.delta.sdk.model.projects.ProjectRef
+import ch.epfl.bluebrain.nexus.delta.sdk.model.projects.ProjectRejection.ProjectAlreadyExists
+import ch.epfl.bluebrain.nexus.delta.sdk.model.projects.ProjectsConfig.AutomaticProvisioningConfig
+import ch.epfl.bluebrain.nexus.delta.sdk.{Acls, Projects}
+import com.typesafe.scalalogging.Logger
+import monix.bio.{IO, UIO}
+
+/**
+  * Automatic project provisioning for users.
+  */
+trait ProjectProvisioning {
+
+  /**
+    * Provision a project for a user
+    *
+    * @param subject a user to provision a project for
+    */
+  def apply(subject: Subject): UIO[Unit]
+
+}
+
+object ProjectProvisioning {
+
+  private val logger: Logger = Logger[ProjectProvisioning]
+
+  /**
+    * Create an instance of [[ProjectProvisioning]]
+    * @param acls               ACLs operations
+    * @param projects           project operations
+    * @param provisioningConfig provisioning configuration
+    */
+  def apply(
+      acls: Acls,
+      projects: Projects,
+      provisioningConfig: AutomaticProvisioningConfig
+  ): ProjectProvisioning = new ProjectProvisioning {
+
+    private def provisionOnNotFound(
+        projectRef: ProjectRef,
+        user: User,
+        acls: Acls,
+        provisioningConfig: AutomaticProvisioningConfig
+    ): UIO[Unit] = {
+      val acl = Acl(AclAddress.Project(projectRef), user -> provisioningConfig.permissions)
+      (for {
+        _ <- acls
+               .append(acl, 0L)(user)
+               .onErrorRecover { case _: AclRejection.IncorrectRev | _: AclRejection.NothingToBeUpdated => () }
+               .mapError { rej => s"Failed to set ACL for user '$user' due to '${rej.reason}'." }
+        _ <- projects
+               .create(
+                 projectRef,
+                 provisioningConfig.fields
+               )(user)
+               .onErrorRecover { case _: ProjectAlreadyExists => () }
+               .mapError { rej => s"Failed to provision project for '$user' due to '${rej.reason}'." }
+      } yield ()).onErrorHandle { err =>
+        logger.warn(err)
+      }
+    }
+
+    override def apply(subject: Subject): UIO[Unit] = subject match {
+      case user @ User(subject, realm) if provisioningConfig.enabled =>
+        (for {
+          org       <- IO.fromOption(provisioningConfig.enabledRealms.get(realm))
+          proj      <- IO.fromEither(Label.apply(subject)).mapError { err =>
```

Comment:
This can really happen ? If yes, we should try to sanitize the subject to fit the constraint (truncate remove some special chars, ...)
Otherwise the user is blocked and won't have its project.

Reply:
It **shouldn't** happen, but  I didn't want to use `Label.unsafe`.  Removing chars could lead to username collisions.

Reply:
We have to sanitize the label to avoid unknown problems. The chance of a collision should be fairly small.